Print to pdf displays navigation bar above pages

ernestlim8 opened the pull request:
Fixes #35 

Navigation bar is not necessary and clutters the pdf pages

Adding a css print query will allow for the pages to be
printed out without showing the navigation bar

Let's specify the behaviour of the media print query
to change the display type of the site-header component,
which contains the navigation bar, to none

We found this fix to be the most suitable as
it only modifies the body content when
printing the document

ernestlim8 commented:
@damithc Apologies for the repeat pull request. I have deleted the branch to edit the commit message. May I enquire on the correct method to edit previous commits without deletion? Thank you.

damithc commented:
> @damithc Apologies for the repeat pull request. I have deleted the branch to edit the commit message. May I enquire on the correct method to edit previous commits without deletion? Thank you.

You can follow the link [here](https://oss-generic.github.io/process/docs/QuestionsIssuesPrs.html#prepare-the-commits-for-publishing) about refactoring commits. Alternatively, can be do the interactive rebase using SourceTree (easier).

After modifying the commits, you should rebase the branch, force-push, and submit a new version using the bot.

damithc commented:
@se-edu/tech-team-level1 any reviews for this PR?
@wxwxwxwx9 any inputs for this PR?

wxwxwxwx9 commented:
@damithc hi prof, I read through the issue thread (#35) and you mentioned that we might want to replace the site navigation with the title of the project. I explored this area and this is the solution that I came up with:

```
@media print {
  /**
  * Prevents page break to cut through content when converting to pdf
  */
  body {
    display: block;
  }
  /**
  * Replaces default site header (with navigation) with a generic header name (AB-3) when converting to pdf. 
  */
  .site-header .wrapper {
    display: none;
  }
  .site-header {
    text-align: center;
  }
  .site-header:before {
    content: "AB-3";
    font-size: 32px;
  }
}
```

<img width="578" alt="Screen Shot 2021-01-11 at 4 07 35 PM" src="https://user-images.githubusercontent.com/37920115/104157922-21a74480-5427-11eb-9b37-c68d0cfd8734.png">
(first page)

![image](https://user-images.githubusercontent.com/37920115/104157949-3a175f00-5427-11eb-8297-1a990d24f501.png)
(second page)

We are only replacing the site header (on the first page) with a project title. So as you can see from second page onwards, there won't be the title.

Is this similar to what you had in mind? 

The original solution proposed by OP works perfectly fine as well, if we don't want a generic project title.

damithc commented:
@wxwxwxwx9 How about we add the project title here in the normal page?
![image](https://user-images.githubusercontent.com/1673303/104159678-be1f1600-542a-11eb-965f-291b7d6d553a.png)

That way, the entire nav bar can be omitted from the printed version and doesn't require students to modify the CSS file to fit their product name.

wxwxwxwx9 commented:
@damithc Ah, I see what you mean. But just to confirm, do you mean that we will always take the project title of the current page? 

For example, if I am at About Us, then I will use the header (which is "About Us") as the title? And the same applies for Developer Guide / About Us. 

Wouldn't that mean we will have 'duplicated' headers? Since "About Us" is already there in the document and we are using it again to replace the site nav.

damithc commented:
> @damithc Ah, I see what you mean. But just to confirm, do you mean that we will always take the project title of the current page?
> 
> For example, if I am at About Us, then I will use the header (which is "About Us") as the title? And the same applies for Developer Guide / About Us.
> 
> Wouldn't that mean we will have 'duplicated' headers? Since "About Us" is already there in the document and we are using it again to replace the site nav.

This is what I had in mind:
* No need to change site nav at all. It can use `AB-3` in every page.
* In [the UserGuide.md](https://github.com/se-edu/addressbook-level3/edit/master/docs/UserGuide.md) frontmatter, change line 3 as `title: "AddressBook Level 3 - User Guide"
* Do the same for the DG page.
* Can do the same for other pages but not really needed as they are not going to be converted to PDF.
* Omit the top nav when printing (as done in this PR). As the UG and DG already contain `AddressBook Level3` in the title, there is no need to 'insert' the project name dynamically using CSS.

Would that work? Note that UG and DG, when viewed on the Web, will have `AB-3` in the top nav and `AddressBook Level 3 - ` in the page title right below the top nav, but I guess we can overlook that slight duplication, for convenience.

ernestlim8 commented:
@damithc I attempted to change the title of the User Guide and Developer Guide. However,  the new extended title also appears in the navigation bar headers. I inspected further and realised that the _config.yml file links the markdown files directly under header_pages in lines 5-7 of the config file. This results in the title being directly transferred over to the navigation bar as well. As such, I do not think that changing the title is a good option.

damithc commented:
> @damithc I attempted to change the title of the User Guide and Developer Guide. However, the new extended title also appears in the navigation bar headers. I inspected further and realised that the _config.yml file links the markdown files directly under header_pages in lines 5-7 of the config file. This results in the title being directly transferred over to the navigation bar as well. As such, I do not think that changing the title is a good option.

I see. Yes, in that case, it's not a good option.
In that case, we can try to solution suggested by @wxwxwxwx9 ?

ernestlim8 commented:
I'm having some trouble with the interactive rebase, as the current master branch is ahead of this branch. The interactive rebase seems to be updating my master branch instead of the 35-omit-navbar-from-pdf from my fork. Please advise on the proper procedure as I am new to rebasing.

damithc commented:
> I'm having some trouble with the interactive rebase, as the current master branch is ahead of this branch. The interactive rebase seems to be updating my master branch instead of the 35-omit-navbar-from-pdf from my fork. Please advise on the proper procedure as I am new to rebasing.

@se-edu/tech-team-level1 can anyone help?

RuiFengg commented:
> I'm having some trouble with the interactive rebase, as the current master branch is ahead of this branch. The interactive rebase seems to be updating my master branch instead of the 35-omit-navbar-from-pdf from my fork. Please advise on the proper procedure as I am new to rebasing.

Since your master branch gets updated, you can checkout to the branch and rebase by doing `git checkout branchname
git rebase master` 
If this doesn't work, try checking out to your branch and then rebasing from there by specifying where you want to rebase from (se-edu/addressbook-level3's master branch). You can check the remote repos you are pushing to by typing `git remote -v`. You should see your own fork eg.   https://github.com/RuiFengg/addressbook-level3.git. But since the master branch of the main repo got updated,  you have to add the base repo  https://github.com/se-edu/addressbook-level3.git  to your remote using `git remote add` naming it whatever you want (your fork's remote is called origin). From then on, in your branch you will be able to pull and rebase from the base repo by specifying the remote name and branch name you're pulling from eg. `git pull --rebase upstream master`. Hope this helps!

damithc commented:
> Since your master branch gets updated, you can checkout to the branch and rebase by doing `git checkout branchname git rebase master`
> If this doesn't work, try checking out to your branch and then rebasing from there by specifying where you want to rebase from (se-edu/addressbook-level3's master branch). You can check the remote repos you are pushing to by typing `git remote -v`. You should see your own fork eg. https://github.com/RuiFengg/addressbook-level3.git. But since the master branch of the main repo got updated, you have to add the base repo https://github.com/se-edu/addressbook-level3.git to your remote using `git remote add` naming it whatever you want (your fork's remote is called origin). From then on, in your branch you will be able to pull and rebase from the base repo by specifying the remote name and branch name you're pulling from eg. `git pull --rebase upstream master`. Hope this helps!

yup, at a higher level, this is what you can do.
1. Checkout the master branch.
2. Pull the upstream master (`git pull upstream master`). That should update your local master.
3. Checkout your pr branch.
4. Rebase the pr branch on master (`git rebase master`)
5. Now, you are ready to do the interactive rebase to refactor commits in your pr branch.

ernestlim8 committed:
Omit the top navigation menu when a page is printed

When a page in the product website (e.g, the User Guide) is converted
into a print view, the top navigation bar appears in printed version.
This is not appropriate as a navigation bar belongs in a Web page but
not in a printed view. This is particularly a problem when documents
in the website are converted into PDF, which is done by 'printing' the
page into a PDF file.

Let's update the behaviour of the media print query to replace the top
navigation menu with simply the project name.

ernestlim8 commented:
@damithc @RuiFengg Thank you for the advice, I managed to understand the rebasing process better with your help. 

The solution suggested by @wxwxwxwx9 is functioning well. Will submit a V2 on the bot.

damithc commented on docs/_sass/minima/_base.scss:
```diff
@@ -271,11 +271,24 @@ table {
- */
 @media print {
+  /**
+  * Prevents page break to cut through content when converting to pdf
```
-> `Prevents page break from cutting through content when converting to PDF`

damithc commented on docs/_sass/minima/_base.scss:
```diff
@@ -271,11 +271,24 @@ table {
     display: block;
   }
+  /**
+  * Replaces default site header (with navigation) with a generic header name (AB-3) when converting to pdf.
```
-> `Replaces the top navigation menu with the project name when converting to PDF`

damithc commented on docs/_sass/minima/_base.scss:
```diff
@@ -271,11 +271,24 @@ table {
+  .site-header:before {
+    content: "AB-3";
+    font-size: 32px;
+  }
```
How about this as the commit message?
```
Omit the top navigation menu when a page is printed

When a page in the product website (e.g, the User Guide) is converted
into a print view, the top navigation bar appears in printed version.
This is not appropriate as a navigation bar belongs in a Web page but
not in a printed view. This is particularly a problem when documents
in the website are converted into PDF, which is done by 'printing' the
page into a PDF file.

Let's update the behaviour of the media print query to replace the top
navigation menu with simply the project name.
```

ernestlim8 replied:
Looks much clearer than my previous one, I'll amend the commit message and comments as stated above.

damithc reviewed (changes requested):
Some cosmetic changes suggested.

damithc commented on docs/_sass/minima/_base.scss:
```diff
@@ -271,11 +271,25 @@ table {
- */
 @media print {
+  /**
+  * Prevents page break from cutting through content when converting to PDF
```
Sorry, I should have noted this earlier. Shall we replace `when converting to PDF` with `when printing` (because the change affects all printing, not just converting to PDF)?

ernestlim8 replied:
I have replaced this comment, and the one on line 282 to `when printing`

damithc commented:
@wxwxwxwx9 for your review (please check the commit message as well).

wxwxwxwx9 reviewed (approved):
Hi Prof, I have looked through the commit message and the code. 

The commit message seems pretty clear and appropriate to me. 

The code and comments LGTM. 

I think we can merge this in.

damithc commented:
Thanks for the fix @ernestlim8 and @wxwxwxwx9 @RuiFengg for the inputs.
